libusb/os/windows_winusb.c

```diff
@@ -2871,6 +2906,228 @@ static int winusbx_submit_bulk_transfer(int sub_api, struct usbi_transfer *itran
+			break;
+		}
+	}
+
+	// Make sure we found the index.
+	if (idx >= interface_desc.bNumEndpoints) {
+		usbi_dbg("Couldn't find the isochronous endpoint %02x.", transfer->endpoint);
+		SetLastError(ERROR_INVALID_PARAMETER);
+		return false;
+	}
+
+	if (IS_XFERIN(transfer)) {
+		// WinUSB only supports isochronous transfers spanning a full USB frames. Later, we might be smarter about this
+		// and allocate a temporary buffer. However, this is harder than it seems as its destruction would depend on overlapped
+		// IO...
+		iso_transfer_size_multiple = (pipe_info_ex.MaximumBytesPerInterval * 8) / pipe_info_ex.Interval;
+		if (transfer->length % iso_transfer_size_multiple != 0) {
+			usbi_dbg("The length of isochronous buffer must be a multiple of the MaximumBytesPerInterval * 8 / Interval");
+			SetLastError(ERROR_INVALID_PARAMETER);
+			return false;
+		}
+	} else {
+		// If this is an OUT transfer, we make sure the isochronous packets are contiguous as this isn't supported otherwise.
+		BOOL size_should_be_zero = FALSE;
+		out_transfer_length = 0;
+		for (idx = 0; idx < transfer->num_iso_packets; ++idx) {
+			struct libusb_iso_packet_descriptor *iso_packet = &transfer->iso_packet_desc[idx];
+			if ((size_should_be_zero && transfer->iso_packet_desc[idx].actual_length != 0) ||
+				(iso_packet->length != transfer->iso_packet_desc[idx].actual_length && idx + 1 < transfer->num_iso_packets && transfer->iso_packet_desc[idx + 1].actual_length > 0)) {
+				SetLastError(ERROR_INVALID_PARAMETER);
+				return false;
+			}
+
+			size_should_be_zero = (transfer->iso_packet_desc[idx].actual_length == 0);
+			out_transfer_length += transfer->iso_packet_desc[idx].actual_length;
+		}
+	}
+
+	windows_try_free_isoch_buffer(itransfer);
+
+	// Register the isochronous buffer to the operating system.
+	ret = WinUSBX[sub_api].RegisterIsochBuffer(pwfd->handle, transfer->endpoint, transfer->buffer, transfer->length, &buffer_handle);
+	if (!ret) {
+		usbi_dbg("RegisterIsochBuffer failed");
+		return false;
+	}
+
+	// Important note: the WinUSB_Read/WriteIsochPipeAsap API requires a ContinueStream parameter that tells whether the isochronous
+	// stream must be continued or if the WinUSB driver can schedule the transfer at its conveniance. Profiling subsequent transfers
+	// with ContinueStream = FALSE showed that 5 frames, i.e. about 5 milliseconds, were left empty between each transfer. This
+	// is critical as this greatly diminish the achievable isochronous bandwidth. We solved the problem using the following strategy:
+	// - Transfers are first scheduled with ContinueStream = TRUE and with winusbx_iso_transfer_continue_stream_callback as user callback.
+	// - If the transfer succeeds, winusbx_iso_transfer_continue_stream_callback restore the user callback and calls its.
+	// - If the transfer fails, winusbx_iso_transfer_continue_stream_callback reschedule the transfer and force ContinueStream = FALSE.
+	if (!transfer_priv->iso_break_stream) {
```

Comment:
It seems like the first request should have ContinueStream be set to FALSE according to https://msdn.microsoft.com/en-us/library/windows/hardware/dn376866(v=vs.85).aspx

At least for a small use case I ran here, setting the first transaction to true resulted in an ERROR_INVALID_PARAMETER error.

Reply:
Let me motivate this one as it has been thought through. The whole thing was thought so that each transfer is first tried with `ContinueStream` set to true. If the transfer fails as a result, the callback detects it and retries it with `ContinueStream` set to false.

Say one schedules N libusb transfers (as this is generally the case with such an async API, e.g. libuvc). The first transfer will fail and be rescheduled but all the subsequent ones will continue the stream and, in this manner, this strategy results in full bandwidth usage. I agree that the first transfer will always fail but so will any transfers that are scheduled too late for the stream to be continued. In some sense, the first transfer is just a special case of not being able to continue a stream and this case is handled in the callback, i.e. `
winusbx_iso_transfer_continue_stream_callback`.
Does this make sense?

Reply:
Yes, but for some reason I kept getting the same error over and over again. Re-tested it now and couldn't reproduce it so might have been something else.

In any case, since the two errors are quite different, and this would mask real invalid parameters, any reason not to set iso_break_stream to true on initialization?

Reply:
Thanks for taking the time to review my code by the way! For this one, I still have concerns. Let me consider two different use cases. Before that let `T` be the time WinUSB waits before scheduling a transfer with `ContinueStream` set to `FALSE`. In my tests, `T` was about 5 milliseconds.

The first use case is the one I mentioned above where one submits `N` transfers and resubmit a transfer in the transfer callback. This is a typical use case for isochronous transfers as it typically consists in periodic transfers and therefore the programmer want to ensure that enough transfer buffers are ready to receive the incoming data. If we set `iso_break_stream` to true on initialization, the first time each of this `N` transfers are scheduled, they are followed by a "blank time" of `T`. In total, this result in `N` times `T` timing overhead at start up. In the current implementation, there is only a timing overhead of `T` for the first transfer.

The second use case I want to consider is the one where `N` is determined at runtime using an adaptive procedure that monitors the current throughput and allocates new transfers if it measures that the bandwidth usage is poor. If we set `iso_break_stream` to true on initialization, each time a new transfer is allocated, this would result in a drop in throughput as it would break the stream.

Does this make sense? What do you think?

Reply:
np, and thanks for your work btw!

I'm not entirely sure when winusb returns this error, but what I had in mind was to make sure ContinueStream==FALSE for the first ever read/write operation on that endpoint. Other operations, even if issued long after, can still have ContinueStream==TRUE and deal with failures accordingly.

That being said, if it doesn't have an ill effects, it's probably not that important. Just resulted in a red herring when debugging other invalid parameter errors :)